accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,321 @@
+All the described operations will be also exposed in an XMLRPC API.
+
+### Import an existing cluster
+
+The requirement is to have an already provisioned management node for the cluster that the user wants to manage with Uyuni/SUSE Manager.
+It is implied that the management node is already capable of reaching the nodes of the cluster (via passwordless SSH or any other mean requested by the cluster provider manager) and all clustering management stack is already installed in the management node.
+
+When the import action is selected, Uyuni/SUSE Manager will register the management node to Uyuni/SUSE Manager (by replicating what is done under Systems > Bootstrapping).
+Additionally, Uyuni/SUSE Manager asks for the name and the type of cluster that the user wants to register.
+
+Uyuni/SUSE Manager will store all entered cluster details data into the database.
+
+After successful registration, the cluster overview will be shown.
+
+### Creating a new cluster
+
+When creating a new cluster, Uyuni/SUSE Manager will ask:
+
+1. Cluster name and cluster type
+2. A minion to provision as the management node identified by a specific System Add-On and provisioned with a Salt Formula that will install the clustering management stack on the minion. A list of available minions will be presented.
+3. Uyuni/SUSE Manager will apply the Salt highstate to provision the system as the clustering management stack.
+
+Only the clustering management stack is provisioned on the management node at this time. At the time of writing, nothing forbids that management stack for all types of a cluster can be installed on the same node.
+
+All other optional actions:
+- provisioning a load balancer
+- copying SSH keys from the management node to aspirant nodes of the cluster
+
+will __not__ be managed by Uyuni/SUSE Manager and must be provisioned manually. In a future follow-up, Uyuni/SUSE Manager can make use of [Salt orchestrate runner](https://docs.saltstack.com/en/latest/topics/orchestrate/orchestrate_runner.html#orchestrate-runner) to provision these complex setups before provisioning the management node.
+
+Uyuni/SUSE Manager will store all entered cluster details data into the database and after successful provisioning of the management node, the cluster overview will be shown.
+
+
+### Cluster overview
+
+Under “Cluster overview”, all clusters known to Uyuni/SUSE Manager will be displayed. When a cluster is selected, a new page listing all the details of the cluster will be shown, containing:
+
+- List of all nodes of the cluster - by invoking the cluster provider manager (via Salt) on the Management Node:
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.list_nodes <cluster name>
+  ```
+    - For CaaSP: `salt <mgmt node> caasp.list_nodes <cluster name>`
+    - For SES: `salt <mgmt node> ses.list_nodes <cluster name>`
+    - For SLE-HA/SHAP: `salt <mgmt node> shap.list_nodes <cluster name>`
```

Comment:
What is the management node manages two types of cluster at the same time: what cluster provider manager would be invoked?

Reply:
@paususe, @mbologna: My original idea was that we can do this just like we do `zypperpkg` vs. the virtual module `pkg`. So the implementation would be called, let's say, `ses.list_nodes`, but you could just call `cluster.list_nodes` because `ses` implements `cluster`. On second thought, unfortunately it's not so easy as long as we aren't running a proxy minion per cluster, because most likely you can only have one virtual implementation at a time. I'd leave this as an exercise for later. As long as the "methods" are all named consistently, it's easy to add the virtual module names later.

Reply:
Yes, that's exactly what I meant: implementing an Uyuni cluster manager should be a matter of implementing an already-defined interface, just like in Java, so all the methods will be named consistently (this can even be lint'ed: "error: your cluster manager does not implement the 'list_nodes' method"), and each cluster manager may also implement other methods as an implementation detail.

Reply:
Agreed, removed the distinction between all cluster providers and that we will use a Salt virtual module if possible.